packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -221,174 +221,164 @@ const runAPI = (plugin, api, args, activity) => {
-        })
-
-        let localReporter = getLocalReporter(activity, reporter)
-
-        localReporter.panicOnBuild({
-          id: `11321`,
-          context: {
-            pluginName,
-            api,
-            message: err instanceof Error ? err.message : err,
-          },
-          error: err instanceof Error ? err : undefined,
-        })
-
-        return null
-      })
-    }).then(results => {
-      if (onAPIRunComplete) {
-        onAPIRunComplete()
-      }
-      // Remove runner instance
-      apisRunningById.delete(apiRunInstance.id)
-      const currentCount = apisRunningByTraceId.get(apiRunInstance.traceId)
-      apisRunningByTraceId.set(apiRunInstance.traceId, currentCount - 1)
+  --apiRunnersActive
+  if (apiRunnersActive === 0) {
+    emitter.emit(`API_RUNNING_QUEUE_EMPTY`)
+  }
 
-      if (apisRunningById.size === 0) {
-        emitter.emit(`API_RUNNING_QUEUE_EMPTY`)
-      }
+  // Filter empty results
+  apiRunInstance.results = results.filter(result => !_.isEmpty(result))
 
-      // Filter empty results
-      apiRunInstance.results = results.filter(result => !_.isEmpty(result))
+  // Filter out empty responses and return if the
+  // api caller isn't waiting for cascading actions to finish.
+  if (!waitForCascadingActions) {
+    apiSpan.finish()
+    resolve(apiRunInstance.results)
+  }
 
-      // Filter out empty responses and return if the
-      // api caller isn't waiting for cascading actions to finish.
-      if (!args.waitForCascadingActions) {
-        apiSpan.finish()
-        resolve(apiRunInstance.results)
-      }
+  // Check if any of our waiters are done.
+  waitingForCasacadeToFinish = waitingForCasacadeToFinish.filter(instance => {
+    // If none of its trace IDs are running, it's done.
+    const apisByTraceIdCount = apisRunningByTraceId.get(instance.traceId)
+    if (apisByTraceIdCount === 0) {
+      instance.span.finish()
+      instance.resolve(instance.results)
+      return false
+    } else {
+      return true
+    }
+  })
 
-      // Check if any of our waiters are done.
-      waitingForCasacadeToFinish = waitingForCasacadeToFinish.filter(
-        instance => {
-          // If none of its trace IDs are running, it's done.
-          const apisByTraceIdCount = apisRunningByTraceId.get(instance.traceId)
-          if (apisByTraceIdCount === 0) {
-            instance.span.finish()
-            instance.resolve(instance.results)
-            return false
-          } else {
-            return true
-          }
-        }
-      )
-      return
+  // This promise will resolve with the results for all the plugin calls for this particular api call.
+  // If args.waitForCascadingActions is true, it will wait for all concurrent calls with the same api to resolve.
+  // If args.waitForCascadingActions is false, it should be resolved by the time the code reaches here
+  return promise
+}
+
+function runPlugin(api, plugin, args, stopQueuedApiRuns, activity, apiSpan) {
```

Comment:
Why not create an async function?

```
async function runPlugin(api, plugin, args, stopQueuedApiRuns, activity, apiSpan) {
  
  try {
    return runAPI(plugin, api, { ...args, parentSpan: apiSpan }, activity))
  } catch(err) {
    let pluginName =
      plugin.name === `default-site-plugin` ? `gatsby-node.js` : plugin.name

    decorateEvent(`BUILD_PANIC`, {
      pluginName: `${plugin.name}@${plugin.version}`,
    })

    let localReporter = getLocalReporter(activity, reporter)

    localReporter.panicOnBuild({
      id: `11321`,
      context: {
        pluginName,
        api,
        message: err instanceof Error ? err.message : err,
      },
      error: err instanceof Error ? err : undefined,
    })

    return null
  }
}

```

Reply:
I had this at some point. It was really nice and clean.

But then I realized removing `waitForCascadingActions` was not backwards compatible. And to make that work you have to juggle a `resolve` function, which means creating your own promise, at which point using `async` is useless and actually hurts performance (because `async` will always generate a promise, even if you just return a value). So there didn't seem to be a point in adding the keyword other than "for the sake of having it".

Reply:
I'm unsure I follow on this comment. You're using new Promise here so we always return a promise? I feel like i'm missing something.

Reply:
What would be the advantage of adding `async` here? You'd still need to create a promise to push onto the stack so another promise can resolve it once the last task completes.

I really don't like this form but right now I did not see a better way out. Within reasonable context :)

Reply:
just making it consistent, nothing much :) I was just wondering why you used async every other place.

Feel free to ignore

Reply:
Ideally I'd just remove the whole `waitForCascadingActions` mechanism. I'm just not convinced it's going to work the same way. If I'm confident enough nothing _actually_ relies on it, I'll remove it and simplify the whole thing again :p